### PR TITLE
 feat(insights): compute insights data in connectInfiniteHits [PART-0]

### PR DESCRIPTION
**Pull request by tkrugg**
**Summary**

This PR adds what's need to add **absolute position** and **queryID** to every hit of connectInfiniteHits

**Result**

Not much to demonstrate yet but from now on every hit will look like this
```js
{
    objectID: "theObjectID",
    __position: 42,
    __queryID: "theQueryID"
}
```

**Global Outline** 
- [x] PART-0 add __positions and __query ID inside connectInfiniteHits [#current](https://github.com/algolia/react-instantsearch/pull/2215)
- [x] PART-1 add __positions and __query ID inside connectHits https://github.com/algolia/react-instantsearch/pull/2216
- [x] PART-2 create connectInsights https://github.com/algolia/react-instantsearch/pull/2217

### CODE REVIEW

File: packages/react-instantsearch-core/src/core/indexUtils.js [hunk 1]
```diff
@@ -6,6 +6,21 @@ export function getIndexId(context) {
     : context.ais.mainTargetedIndex;
 }
 
+export function addAbsolutePositions(hits, hitsPerPage, page) {
```
Comment (samouss) [inline packages/react-instantsearch-core/src/core/indexUtils.js]: It should be more inside `utils` rather than `indexUtils`. This file mainly deals with `context` and `searchState`.

File: packages/react-instantsearch-core/src/connectors/connectInfiniteHits.js [hunk 2]
```diff
@@ -53,7 +55,10 @@ export default createConnector({
       };
     }
 
-    const { hits, page, nbPages } = results;
+    const { page, nbPages } = results;
+
+    let hits = addAbsolutePositions(results.hits, results.hitsPerPage, page);
```
Comment (samouss) [inline packages/react-instantsearch-core/src/connectors/connectInfiniteHits.js]: I would rather use two different variables to actually inform that they've changed e.g. `hitsWithPositions`.

File: packages/react-instantsearch-core/src/core/indexUtils.js (hunk 1 above)
Comment (tkrugg) [reply packages/react-instantsearch-core/src/core/indexUtils.js]: Thanks for advising on that, I initially has a FIXME: find theses functions a better place but linting got to remove it :)
Will do

File: packages/react-instantsearch-core/src/connectors/connectInfiniteHits.js (hunk 2 above)
Comment (tkrugg) [reply packages/react-instantsearch-core/src/connectors/connectInfiniteHits.js]: I agree, let do it!

File: packages/react-instantsearch-core/src/core/utils.js [hunk 3]
```diff
@@ -44,3 +44,18 @@ export const removeEmptyKey = obj => {
...
+
+export function addAbsolutePositions(hits, hitsPerPage, page) {
+  return hits.map((hit, index) => ({
+    ...hit,
+    __position: hitsPerPage * page + index + 1,
+  }));
+}
+
+export function addQueryID(hits, queryID) {
+  if (!queryID) return hits;
```
Comment (francoischalifour) [inline packages/react-instantsearch-core/src/core/utils.js]: Can we avoid inline returns?

Comment (tkrugg) [reply packages/react-instantsearch-core/src/core/utils.js]: I know this is a pretty common coding style preference but it doesn't seem we follow it in the codebase: 
eg. https://github.com/algolia/react-instantsearch/blob/ec9e0fbd6106d1c3e47f1dbfa6eaac3e20af6bd5/packages/react-instantsearch-core/src/components/Index.js#L82

If we decide this is a coding style we want to adopt, we should just add it to linting. 
https://eslint.org/docs/rules/curly#require-following-curly-brace-conventions-curly

Comment (francoischalifour) [reply packages/react-instantsearch-core/src/core/utils.js]: It's mentioned in our [ESLint config](https://github.com/algolia/eslint-config-algolia/blob/e6661db323a69d4304a4bc84571e2263e572a6a4/rules/base.js#L45) but for some reason not picked up.

We can still apply this rule for new code.

File: packages/react-instantsearch-core/src/connectors/connectInfiniteHits.js (hunk 2 above)
Comment (samouss) [reply packages/react-instantsearch-core/src/connectors/connectInfiniteHits.js]: Yes.

File: packages/react-instantsearch-core/src/connectors/__tests__/connectInfiniteHits.js [hunk 4]
```diff
@@ -59,10 +64,25 @@ describe('connectInfiniteHits', () => {
...
+          page: 0,
+          hitsPerPage: 6,
+          nbPages: 10,
+          queryID: 'theQueryID_0',
+        },
       });
 
-      expect(res1.hits).toEqual(hits);
+      expect(res1.hits).toEqual(hits.map(hit => expect.objectContaining(hit)));
+      expect(res1.hits.map(hit => hit.__position)).toEqual([1, 2, 3, 4, 5, 6]);
```
Comment (samouss) [inline packages/react-instantsearch-core/src/connectors/__tests__/connectInfiniteHits.js]: `_highlightResults` contains only one `_` but our `__position` contains two `__`. Is it on purpose?

File: packages/react-instantsearch-core/src/core/indexUtils.js [hunk 5]
```diff
@@ -1,4 +1,4 @@
-import { has, omit, get } from 'lodash';
+import { get, has, omit } from 'lodash';
```
Comment (samouss) [inline packages/react-instantsearch-core/src/core/indexUtils.js]: We can revert this change, unless you really care about it.

File: packages/react-instantsearch-core/src/connectors/__tests__/connectInfiniteHits.js (hunk 4 above)
Comment (tkrugg) [reply packages/react-instantsearch-core/src/connectors/__tests__/connectInfiniteHits.js]: I had `_` initially in the RFC and someone mentioned that we only `__` is considered truly private
hence having `__hitIndex` currently, rather than `_hitIndex`.
Ready to change it if necessary (here and in vanilla)

File: packages/react-instantsearch-core/src/core/indexUtils.js (hunk 5 above)
Comment (tkrugg) [reply packages/react-instantsearch-core/src/core/indexUtils.js]: I don't, I was sure it was prettier at first I can't find a rule that does that. Reverting

File: packages/react-instantsearch-core/src/connectors/__tests__/connectInfiniteHits.js (hunk 4 above)
Comment (samouss) [reply packages/react-instantsearch-core/src/connectors/__tests__/connectInfiniteHits.js]: You're right it's probably better at least the API won't override it by mistake.

File: packages/react-instantsearch-core/src/core/utils.js (hunk 3 above)
Comment (tkrugg) [reply packages/react-instantsearch-core/src/core/utils.js]: of course, but it's best to leverage the tooling to move the focus away from style in reviews.

I looked a bit and it looks like it's disabled by prettier which we use in algolia/react
https://github.com/prettier/eslint-config-prettier/blob/master/index.js#L9

We should reenable it if we care about this curly rule

Comment (tkrugg) [reply packages/react-instantsearch-core/src/core/utils.js]: this is fixed btw, but looking forward to continue the discussion on adding this rule to our linter.

File: packages/react-instantsearch-core/src/connectors/__tests__/connectInfiniteHits.js [hunk 6]
```diff
@@ -21,7 +21,10 @@ describe('connectInfiniteHits', () => {
         results: { hits, page: 0, hitsPerPage: 2, nbPages: 3 },
       });
 
-      expect(props).toEqual({ hits, hasMore: true });
+      expect(props).toEqual({
+        hits: hits.map(hit => expect.objectContaining(hit)),
```
Comment (Haroenv) [inline packages/react-instantsearch-core/src/connectors/__tests__/connectInfiniteHits.js]: same comment as for Hits, this could be better something more precise

File: packages/react-instantsearch-core/src/connectors/__tests__/connectInfiniteHits.js [hunk 7]
```diff
@@ -83,10 +138,63 @@ describe('connectInfiniteHits', () => {
...
+        5,
+        6,
+        // page: 1, hitsPerPage: 6
+        7,
+        8,
+        9,
+        10,
+        11,
+        12,
+        // page: 2, hitsPerPage: 8
```
Comment (Haroenv) [inline packages/react-instantsearch-core/src/connectors/__tests__/connectInfiniteHits.js]: I'd clarify after this that it's an existing limitation of the engine of not being able to correctly compute an offset

Comment (tkrugg) [reply packages/react-instantsearch-core/src/connectors/__tests__/connectInfiniteHits.js]: yes, I'll add the following
> hitsPerPage changed from 6 to 8, elements 13-16 are skipped

File: packages/react-instantsearch-core/src/connectors/__tests__/connectInfiniteHits.js (hunk 6 above)
Comment (tkrugg) [reply packages/react-instantsearch-core/src/connectors/__tests__/connectInfiniteHits.js]: I added because now hits get `__position` and `__queryID`.
This tests is strictly about checking hits identity

File: packages/react-instantsearch-core/src/connectors/__tests__/connectInfiniteHits.js (hunk 7 above)
Comment (tkrugg) [reply packages/react-instantsearch-core/src/connectors/__tests__/connectInfiniteHits.js]: fixed here: https://github.com/algolia/react-instantsearch/pull/2217/commits/45220e54ada5f9734405e7b0d971dc6f238243ee

File: packages/react-instantsearch-core/src/connectors/__tests__/connectInfiniteHits.js (hunk 6 above)
Comment (Haroenv) [reply packages/react-instantsearch-core/src/connectors/__tests__/connectInfiniteHits.js]: `hits.map(hit => expect.objectContaining(hit))` will always pass, except if hit isn't an object though. I don't see how this asserts anything.

File: packages/react-instantsearch-core/src/connectors/connectInfiniteHits.js [hunk 8]
```diff
@@ -53,14 +54,24 @@ export default createConnector({
       };
     }
 
-    const { hits, page, nbPages } = results;
+    const { page, nbPages } = results;
+
+    const hitsWithPositions = addAbsolutePositions(
+      results.hits,
+      results.hitsPerPage,
```
Comment (samouss) [inline packages/react-instantsearch-core/src/connectors/connectInfiniteHits.js]: We should stay consistent either we destruct either we use the dot form.

Comment (tkrugg) [reply packages/react-instantsearch-core/src/connectors/connectInfiniteHits.js]: fixed here 52ebc802201bbbbf7a6dbb3c719802bd1347c8ad
